File: frontend/src/services/signalingService.js
/**
 * signalingService.js
 *
 * Manages the STOMP over SockJS connection to the Spring Boot backend.
 * The backend is the sole authority on routing — this service is a thin transport layer.
 */

import { Client } from '@stomp/stompjs';
import SockJS from 'sockjs-client';

const WS_URL = '/ws'; //hna

let stompClient = null;
let stompSessionId = null;
let onSignalCallback = null;
let onReconnectCallback = null;

function extractSockJsSessionId(ws) {
  try {
    // 1. Try the underlying native WebSocket URL first (most specific, contains session ID)
    let url = ws?._transport?.ws?.url || ws?.conn?._transport?.ws?.url;
    // 2. Fallback to other transport properties
    if (!url || url.endsWith('/ws')) {
      url = ws?._transport?.url || ws?.conn?._transport?.url || ws?.url;
    }

    if (!url) return null;
    console.log('[STOMP] Found WebSocket URL:', url);
    const segments = url.split('/');
    if (segments.length >= 2) {
      const last = segments[segments.length - 1];
      const secondToLast = segments[segments.length - 2];
      
      // SockJS format: ws://<host>/ws/<server>/<session>/websocket
      if (last === 'websocket' || last === 'xhr' || last === 'xhr_streaming') {
        return secondToLast;
      }
      
      // General format: search for the 'ws' segment
      const wsIndex = segments.indexOf('ws');
      if (wsIndex !== -1 && segments.length > wsIndex + 2) {
        return segments[wsIndex + 2];
      }
    }
  } catch (e) {
    console.error('[STOMP] Error parsing WebSocket URL:', e);
  }
  return null;
}

/**
 * Connect to the backend WebSocket broker.
 * @param {function} onSignal - called with each SignalingMessage from server
 * @param {function} onReconnect - called with new stompSessionId when STOMP reconnects
 * @returns {Promise<string>} resolves with the STOMP session ID
 */
export function connect(onSignal, onReconnect) {
  return new Promise((resolve, reject) => {
    onSignalCallback = onSignal;
    onReconnectCallback = onReconnect;
    let isFirstConnect = true;

    stompClient = new Client({
      webSocketFactory: () => new SockJS(WS_URL, null, { transports: ['websocket'] }),
      reconnectDelay: 3000,
      onConnect: (frame) => {
        console.log('[STOMP] CONNECTED frame headers:', JSON.stringify(frame.headers));
        
        // Extract session ID directly from the active WebSocket connection URL, 
        // fallback to STOMP headers if needed.
        const wsSessionId = extractSockJsSessionId(stompClient.webSocket);
        stompSessionId = wsSessionId || frame.headers['user-name'] || frame.headers['session'];
        
        console.log('[STOMP] Subscribing with session ID:', stompSessionId);

        // Subscribe to our personal queue (session-specific)
        // The server sends signals to /queue/signal-{stompSessionId}
        stompClient.subscribe(`/queue/signal-${stompSessionId}`, (msg) => {
          console.log('[STOMP] Received message:', msg.body);
          try {
            const signal = JSON.parse(msg.body);
            if (onSignalCallback) onSignalCallback(signal);
          } catch (e) {
            console.error('Failed to parse signal:', e);
          }
        });

        if (isFirstConnect) {
          isFirstConnect = false;
          resolve(stompSessionId);
        } else {
          // Reconnected with a new stompSessionId — notify caller to re-join queue
          console.log('STOMP reconnected with new session:', stompSessionId);
          if (onReconnectCallback) onReconnectCallback(stompSessionId);
        }
      },
      onStompError: (frame) => {
        console.error('STOMP error', frame);
        reject(new Error(frame.headers?.message || 'WebSocket error'));
      },
      onDisconnect: () => {
        console.log('[STOMP] disconnected');
      },
    });

    stompClient.activate();
  });
}

/**
 * Send a signaling message to the backend.
 * @param {object} message - SignalingMessage payload
 */
export function sendSignal(message) {
  if (!stompClient || !stompClient.connected) {
    console.warn('STOMP not connected — cannot send signal');
    return;
  }
  stompClient.publish({
    destination: '/app/signal',
    body: JSON.stringify(message),
  });
}

/**
 * Request to join the matchmaking queue.
 * @param {string} sessionToken - backend session token
 */
export function joinQueue(sessionToken) {
  if (!stompClient || !stompClient.connected) return;
  stompClient.publish({
    destination: '/app/queue',
    body: JSON.stringify({
      type: 'SEARCHING',
      senderSessionToken: sessionToken,
    }),
  });
}

/** Disconnect from the broker */
export function disconnect() {
  if (stompClient) {
    stompClient.deactivate();
    stompClient = null;
  }
}

export function getStompSessionId() {
  return stompSessionId;
}

File: mobile/lib/services/signaling_service.dart
// lib/services/signaling_service.dart
import 'dart:convert';
import 'package:stomp_dart_client/stomp_dart_client.dart';
import '../config/app_config.dart';
import '../models/models.dart';

typedef SignalHandler = void Function(SignalingMessage message);

class SignalingService {
  late StompClient _client;
  String? _stompSessionId;
  SignalHandler? _onSignal;
  bool _connected = false;

  String? get stompSessionId => _stompSessionId;
  bool get isConnected => _connected;

  /// Connects to the backend STOMP broker and subscribes to personal queue.
  Future<void> connect({required SignalHandler onSignal}) async {
    _onSignal = onSignal;

    final completer = Future<void>.value();
    bool resolved = false;

    _client = StompClient(
      config: StompConfig.sockJS(
        url: AppConfig.wsUrl,
        onConnect: (frame) {
          _connected = true;
          // Extract session ID from STOMP frame headers
          _stompSessionId = frame.headers['user-name'] ??
              frame.headers['session'] ??
              DateTime.now().millisecondsSinceEpoch.toString();

          // Subscribe to personal signal queue
          _client.subscribe(
            destination: '/queue/signal-$_stompSessionId',
            callback: (frame) {
              if (frame.body == null) return;
              try {
                final json = jsonDecode(frame.body!) as Map<String, dynamic>;
                final msg = SignalingMessage.fromJson(json);
                _onSignal?.call(msg);
              } catch (e) {
                print('Signal parse error: $e');
              }
            },
          );

          if (!resolved) resolved = true;
        },
        onStompError: (frame) {
          print('STOMP error: ${frame.body}');
          _connected = false;
        },
        onDisconnect: (_) {
          _connected = false;
        },
        onWebSocketError: (error) {
          print('WS error: $error');
          _connected = false;
        },
        reconnectDelay: const Duration(seconds: 3),
      ),
    );

    _client.activate();

    // Wait for connection (simple polling approach)
    for (int i = 0; i < 30; i++) {
      await Future.delayed(const Duration(milliseconds: 200));
      if (_connected) return;
    }
    throw Exception('WebSocket connection timed out');
  }

  /// Sends a signaling message to the backend for routing.
  void send(SignalingMessage message) {
    if (!_connected) {
      print('Cannot send — not connected');
      return;
    }
    _client.send(
      destination: '/app/signal',
      body: jsonEncode(message.toJson()),
    );
  }

  /// Joins the matchmaking queue.
  void joinQueue(String sessionToken) {
    if (!_connected) return;
    _client.send(
      destination: '/app/queue',
      body: jsonEncode({
        'type': 'SEARCHING',
        'senderSessionToken': sessionToken,
      }),
    );
  }

  void disconnect() {
    _client.deactivate();
    _connected = false;
  }
}
